libyul/ASTNodeRegistry.h

```diff
@@ -0,0 +1,94 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace solidity::yul
+{
+
+/// Instances of the `ASTNodeRegistry` are immutable containers describing a labelling of nodes inside the AST.
+/// Each element of the AST that possesses a label has a `ASTNodeRegistry::NodeId`, with which the label can
+/// be queried in O(1).
+/// Preferred way of creating instances is via `ASTNodeRegistryBuilder` when parsing/importing and
+/// via `NodeIdDispenser` during/after optimization.
+class ASTNodeRegistry
+{
+public:
+	/// unsafe to use from a different registry instance, it is up to the user to safeguard against this
+	using NodeId = size_t;
```

Comment:
I think it should be `ID` for consistency with other acronyms like `AST`:

```suggestion
	using NodeID = size_t;
```
Here and generally in all the other names.

I'm also still not sure `NodeID` and `ASTNodeRegistry` really reflect the purpose well. I see that in AST you even call the registry `labels`. Wouldn't `LabelID` and `ASTLabelRegistry` fit better? I don't think we're really using it as a node ID. We have some node types without any and technically we could also define some nodes where having more than one would make sense.

Reply:
Yes you are right. Especially since same labels will have the same id and before disambiguation it is not necessarily the case that the ids are disjoint in the tree. I have been wondering if it wouldn't be better to do this as actual node ids but then one would have to think about scoping and such.